### PR TITLE
Introduce REST test transformations

### PR DESCRIPTION
**Pull request by jakelandis**
This commit introduces the ability to programmatically transform YAML based REST tests. Specifically this initial commit introduces the ability to inject HTTP headers into the YAML REST tests.  Additional capabilities (transforms) will be introduced in future commits. 

Transforming REST tests is a key component to test REST API compatibility.  Eventually (not included here) the ability to transform tests will be integrated with Gradle such that the tests that are transformed prior to being executed. This commit does not have any REST API compatibility specific changes, and only introduces the code that will be used for that testing. 

Transforming REST tests at the YAML level was chosen over updating the test runners to ensure that these tests may (at a future time) be used by the various clients that re-use the YAML based REST tests. By transforming the tests themselves, there is a consistent view across all runners for exactly how and what to test when running N-1 tests against a N cluster with REST compatibility. 

----

To get an idea of how this will be used:
https://github.com/jakelandis/elasticsearch/blob/compat_header_inject_with_gradle/modules/ingest-common/build.gradle#L20
https://github.com/jakelandis/elasticsearch/blob/compat_header_inject_with_gradle/buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/YamlRestCompatTestPlugin.java#L148
https://github.com/jakelandis/elasticsearch/blob/compat_header_inject_with_gradle/buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/RestCompatTestTransformTask.java

----

I also added "humanDebug" flag to the tests that will output the transformed tests to stdout in case you want to run the tests and see the output directly.

### CODE REVIEW

**Comment by jakelandis** [comment]
@elasticmachine update branch

**Comment by nik9000** [comment]
This seems like a very reasonable way to do it to me. I'd love to be able to `cat` the transformed yml files and this seems like a very reasonable way to get that.

**Review by nik9000 — APPROVED** [review]
Given that @mark-vieira is going to be pretty busy I think we can get this in now and beg forgiveness later if we find something untoward. It makes sense to me. We'll get a closer review on the actual build changes required when we build the files and such.

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransform.java
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransform.java:27]: Is there a reason we can't merge this interface with `RestTestTransformGlobalSetup` and `RestTestTransformGlobalTeardown`. Implementations need not actually implement those methods as we could provide default noop implementations in the interface itself or in an abstract base class.

Doing so would eliminate all the nasty `instanceof` runtime checks in `RestTestTransformer`.

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java:110]: This seems like something that should exist in the transform, as again, we are now coupled to the corresponding interface via an `instanceof` check. Perhaps put this in an abstract implementation of `RestTestTransform` `transformTest` method and have `InjectHeaders` extend that.

File: buildSrc/src/test/java/org/elasticsearch/gradle/test/rest/transform/InjectHeaderTests.java
Comment (mark-vieira) [inline buildSrc/src/test/java/org/elasticsearch/gradle/test/rest/transform/InjectHeaderTests.java:46]: Boy, I wish we could start moving some of these unit tests to spock as well but that's outside the scope of this PR.

**Review by mark-vieira — APPROVED** [review]
I left a few comments, otherwise I'll defer to someone that's more familiar with the YAML test stuff to review.

We'll definitely want to chat about how to do the build integration stuff. Effectively, the transforms themselves will be an input to these tasks so we need to track those implementations as inputs, which can be tricky depending on where those implementations live.

**Comment by mark-vieira** [comment]
@breskeby heads up on this as well. This is very reminiscent of what I did for the Gradle documentation stuff :wink:. I remember making cacheability work properly on that was a non-trivial effort and we'll want to ensure the same thing here.

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java
Comment (pgomulka) [inline buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java:70]: this could be done before the `tests` for loop?

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java
Comment (pgomulka) [inline buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java:84]: can you put a comment about this expression?
Is the intention that if there is no setupSection, but there is a transformation for a setup, then add a setup?
but it also looks like only one can be added? What if I have two setup transformations ?

**Review by pgomulka — APPROVED** [review]
I left two minor comments
LGTM

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java
Comment (jakelandis) [reply buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java:70]: indeed it can, nice catch.

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java
Comment (jakelandis) [reply buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java:84]: > can you put a comment about this expression?
will do. 

> Is the intention that if there is no setupSection, but there is a transformation for a setup, then add a setup?
yes, that is the intention (create one if missing)

> but it also looks like only one can be added? What if I have two setup transformations ?
good catch . There is a bug as-is with multiple setup/teardown transforms which will get fixed ... however since there is only at most 1 with this PR it is not tested yet , but will likely run into that sooner then later once i introduce more transforms.

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java
Comment (jakelandis) [reply buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java:115]: The `RestTestTransformer` operates against the whole YAML test, and the implementations of `RestTestTransform` operate only on the specific node passed (from one or multiple tests) that it passed. This is to help minimize the number of times we need to traverse the JSON tree. By keeping the JSON tree traversal here (as opposed to the individual transform) allows for multiple transforms with a single traversal for all transforms that are initiated by a named object key.  Currently it is only looks for "do" to inject headers, but future transforms will will be able to implement that interface with a different object key name and do something different like replace the value of object.  Further, this method will likely get extended and renamed in future iterations to allow for more traits that kick off a transform (i.e. replacing one text value with another) while keeping the JSON traversals needed at a minimum.

**Comment by jakelandis** [comment]
@elasticmachine update branch

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java:115]: Understood. Does `JsonNode` not contain a reference to it's parent such that the transform itself could make the deterinmation it's "inside" a `do:` block?

Comment (jakelandis) [reply buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java:115]: It does not maintain a parent reference, however, with the recursive approach here it is pretty simple to carry forward the parent (which is only ever an object or an array which kicks off the recursion). 

Not sure that would help however, the goal here is to find all objects with the key "do" once we find that then we can transform it. In this specific case we only need the object itself, in other (future) cases we will need to pass the parent not the thing found. For example if we need an AddSyblingObject then we would need to pass in the parent of found object. 

Perhaps I could change the name of the method to avoid confusion... the method should probably be called `traverseTest`.

**Comment by jakelandis** [comment]
Thanks everyone for the reviews !

**Pull request merged**
